include/pybind11/pybind11.h

```diff
@@ -906,13 +906,40 @@ void call_operator_delete(T *p) { T::operator delete(p); }
 
 inline void call_operator_delete(void *p) { ::operator delete(p); }
 
+// Wrapper around cpp_function that binds a method pointer of a base class with a derived class
+// argument, as opposed to calling cpp_function directly which always uses the inferred class.  For
+// any other type of argument, this is identical to cpp_function.
+template <typename Derived, typename Func, typename SFINAE = void> struct method_adaptor_impl {
+    template <typename... Args> static cpp_function adapt(Args &&... args) { return {std::forward<Args>(args)...}; }
+};
+template <typename Derived, typename Return, typename Class, typename... Arg>
+struct method_adaptor_impl<Derived, Return (Class::*)(Arg...), enable_if_t<is_strict_base_of<Class, Derived>::value>> {
+    template <typename... Extra> static cpp_function adapt(Return (Class::*f)(Arg...), const Extra&... extra) {
+        return {[f](Derived *c, Arg... args) -> Return { return (c->*f)(args...); }, extra...};
```

Comment:
There's no advantage to forwarding extras given the way they are used.  The generic one uses perfect forwarding because there could be an potential advantage in forwarding a stateful lambda (the first argument), and it's simpler to just forward everything.